pkg/master/controller.go

```diff
@@ -284,146 +285,154 @@ type EndpointReconciler interface {
 	// managed only by ReconcileEndpoints; therefore, to understand this, you need only
 	// understand the requirements.
 	//
+	// It uses auxilary config map to keep expiration times for apiservers.
+	// Each apiserver separately update its expiration time and removes all expired apiservers.
+	//
 	// Requirements:
 	//  * All apiservers MUST use the same ports for their {rw, ro} services.
 	//  * All apiservers MUST use ReconcileEndpoints and only ReconcileEndpoints to manage the
 	//      endpoints for their {rw, ro} services.
 	//  * ReconcileEndpoints is called periodically from all apiservers.
-	ReconcileEndpoints(serviceName string, ip net.IP, endpointPorts []api.EndpointPort, reconcilePorts bool) error
+	ReconcileEndpoints(configmapName, serviceName string, ip net.IP, endpointPorts []api.EndpointPort, reconcilePorts bool, now time.Time) error
 }
 
-// masterCountEndpointReconciler reconciles endpoints based on a specified expected number of
-// masters. masterCountEndpointReconciler implements EndpointReconciler.
-type masterCountEndpointReconciler struct {
-	masterCount    int
-	endpointClient coreclient.EndpointsGetter
+// dynamicEndpointReconciler reconciles endpoints based on a config map with expiration times.
+// dynamicEndpointReconciler implements EndpointReconciler.
+type dynamicEndpointReconciler struct {
+	endpointClient  coreclient.EndpointsGetter
+	configmapClient coreclient.ConfigMapsGetter
 }
 
-var _ EndpointReconciler = &masterCountEndpointReconciler{}
+var _ EndpointReconciler = &dynamicEndpointReconciler{}
+
+// NewDynamicEndpointReconciler creates a new EndpointReconciler that reconciles based on an auxiliary config map.
+func NewDynamicEndpointReconciler(endpointClient coreclient.EndpointsGetter,
+	configmapClient coreclient.ConfigMapsGetter) *dynamicEndpointReconciler {
+	return &dynamicEndpointReconciler{
+		endpointClient:  endpointClient,
+		configmapClient: configmapClient,
+	}
+}
 
-// NewMasterCountEndpointReconciler creates a new EndpointReconciler that reconciles based on a
-// specified expected number of masters.
-func NewMasterCountEndpointReconciler(masterCount int, endpointClient coreclient.EndpointsGetter) *masterCountEndpointReconciler {
-	return &masterCountEndpointReconciler{
-		masterCount:    masterCount,
-		endpointClient: endpointClient,
+// updateCM updates TTL configmap for API servers: renews TTL for key server and removes all expired entries.
+func updateCM(key string, m *api.ConfigMap, now time.Time) {
+	const format = "20060102 150405 MST"
+	window := 3 * DefaultEndpointReconcilerInterval
+	newData := map[string]string{}
+	for k, v := range m.Data {
+		ttl, err := time.Parse(format, v)
+		if err != nil {
+			continue
+		}
+		if now.Before(ttl) {
```

Comment:
Isn't this unstable if clocks drift?  The advantage of etcd TTL is that the master clock is all that matters, but here if the apiservers drift out of the window they will be omitted from the quorum and not fixed.

Scenario:

1. Three masters
2. Master 1 clock drifts out of window - it flaps in and out of the cluster
3. Master 2 clock drifts out of window - it flaps in and out of the cluster
3. Master 1 and 2 are actively trying to kick 3 out of the cluster at this point, but depending on who wins the race clients will flip between the two sets (1-2 and 3)

I think if we go down this route (after looking through the discussion from the other issue) then this needs to account for clock drift as observed by other clients.    It's not enough for this to assume the masters time stays in sync throughout the entire time the masters are up (because the masters don't converge on a clock).